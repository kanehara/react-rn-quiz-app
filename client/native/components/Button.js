import styled from 'styled-components'
import {View} from 'react-native'
import Text from 'native/components/Text'
import React from 'react'

const TouchableHighlight = styled.TouchableHighlight`
  background-color: ${({ theme }) => theme.buttonColor};
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  
  width: 100%;
  align-items: center;
  justify-content: center;
  padding: 5px 10px;
  margin: 10px;
`

const ButtonText = styled(Text)`
  font-weight: 700;
`

const Button = props => (
  <TouchableHighlight 
    style={props.style} 
    onPress={props.onPress}
    underlayColor='rgba(0, 0, 0, 0.15)'
  >
    <View>
      <ButtonText>{props.title}</ButtonText>
    </View>
  </TouchableHighlight>
)

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export default Button
