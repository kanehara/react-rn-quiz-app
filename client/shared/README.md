# The shared module for the trivia clients
A module housing common JS code between the web and React Native clients.

## Contains
* Redux actions, reducers, getters
* Apollo client used in redux layer to fetch data
* styled components theme