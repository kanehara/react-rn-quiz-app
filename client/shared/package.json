{
  "name": "shared",
  "version": "0.0.1",
  "scripts": {
    "test": "jest"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "babel-jest": "^24.0.0",
    "jest": "^24.0.0",
    "unfetch": "^4.0.1"
  }
}
