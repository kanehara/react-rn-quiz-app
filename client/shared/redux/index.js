import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { quizReducer } from './quiz'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ 
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) 
    : compose

const persistedReducer = persistReducer({
  key: 'root',
  storage
}, quizReducer)

const store = createStore(
  persistedReducer,
  composeEnhancers(
    applyMiddleware(thunk)
  )
)

const persistor = persistStore(store)

export {
  store,
  persistor
}
