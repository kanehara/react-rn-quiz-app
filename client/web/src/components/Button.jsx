import styled from 'styled-components'

const Button = styled.button`
  padding: 10px;
  width: 100%;
  font-size: 18px;
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  font-weight: bold;
  background-color: ${props => props.theme.buttonColor};

  &:hover {
    cursor: pointer;
  }
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export default Button
