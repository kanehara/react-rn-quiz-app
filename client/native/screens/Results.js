import React from 'react'
import {Button, ScrollView} from 'react-native'
import Text from 'native/components/Text'
import ScreenView from './ScreenView'
import { connect } from 'react-redux'
import { getters, actions } from 'shared/redux/quiz'
import styled from 'styled-components'
import {H1, H2} from 'native/components/Header'

const Question = ({question, answer}) => (
  <React.Fragment>
    <Text>
      {question.correct_answer === answer ? 'O ' : 'X '} 
      {question.question} 
      - {question.correct_answer}
    </Text>
  </React.Fragment>
)

class Results extends React.Component {
  componentDidMount () {
    if (this.props.inProgress) {
      this.props.navigation.navigate('quiz')
    }
  }

  reset = () => {
    this.props.reset()
    this.props.navigation.navigate('home')
  }

  render() {
    return (
      <ScreenView>
        <ScrollView>
          <H1>Results</H1>
          <H2>{this.props.score} / {this.props.questions.length}</H2>
          {this.props.questions.map((q, i) => 
            <Question key={i} question={q} answer={this.props.answers[i]} />
          )}
        </ScrollView>
        <Button onPress={this.reset} title='Restart' />
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    score: getters.score(state),
    finished: getters.finished(state),
    inProgress: getters.inProgress(state),
    answers: state.answers,
    questions: state.questions
  }),
  dispatch => ({
    reset: () => dispatch(actions.reset())
  })
)(Results)
