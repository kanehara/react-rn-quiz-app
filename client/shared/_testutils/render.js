import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

