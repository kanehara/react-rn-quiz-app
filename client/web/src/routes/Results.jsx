import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import styled, {css} from 'styled-components'
import Button from '../components/Button'
import {reset} from 'shared/redux/actions'
import {Flex} from 'rayout'

const QuestionHeader = styled.h3`
  color: ${({ correct }) => correct ? 'green' : 'red'};
  margin-bottom: 5px;
  cursor: pointer;
`

const QuestionDetails = styled.div`
  height: 0;
  display: none;
  opacity: 0;
  transition: height 0.5s;
  transition: opacity 0.5s;

  ${({show}) => show && css`
    display: block;
    height: auto;
    opacity: 1;
  `}
`

class Question extends React.Component {
  state = {
    showDetails: false
  }

  render () {
    const {question} = this.props
    return (
      <Flex flexDirection='column'>
        <QuestionHeader 
          onClick={() => this.setState({ showDetails: !this.state.showDetails })}
          correct={question.answer === question.correct_answer}
        >
          {question.question}
        </QuestionHeader>
        <QuestionDetails show={this.state.showDetails}>
          <p>You answered: {question.answer}</p>
          <p>Correct answer: {question.correct_answer}</p>
        </QuestionDetails>
      </Flex>
    )
  }
}

class Results extends React.Component {
  get score () {
    return this.props.questions.reduce((prev, curr) => {
      return prev + (curr.answer === curr.correct_answer ? 1 : 0)
    }, 0)
  }

  restart = () => {
    this.props.reset()
    this.props.history.replace('/')
  }

  render () {
    return this.props.finished
      ? (
        <div>
          <h2>You scored</h2>
          <h2>{this.score} / {this.props.questions.length}</h2>
          <Button onClick={this.restart}>Restart</Button>
          {this.props.questions.map((q, i) => (
            <Question question={q} key={i} />
          ))}
        </div>
      )
      : (
        <Redirect to='/' />
      )
  }
}

export default connect(
  ({ finished, questions }) => ({
    questions,
    finished
  }),
  dispatch => ({
    reset: () => dispatch(reset())
  })
)(Results)
