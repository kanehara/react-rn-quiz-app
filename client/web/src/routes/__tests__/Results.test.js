import {Results} from '../Results'
import React from 'react'
import { render } from 'shared/_testutils'
import {MemoryRouter} from 'react-router-dom'
import Button from 'web/src/components/Button'

describe('Results', () => {
  const getMockHistory = () => ({
    replace: jest.fn()
  })

  it('redirects to / if not finished', () => {
    expect(render(
      <MemoryRouter>
        <Results finished={false} />
      </MemoryRouter>
    )).toMatchSnapshot()
  })
  
  it('redirects to /quiz if inProgress', () => {
    expect(render(
      <MemoryRouter>
        <Results inProgress />
      </MemoryRouter>
    )).toMatchSnapshot()
  })
  
  it('renders results', () => {
    expect(render(
      <MemoryRouter>
        <Results 
          questions={[
            {correct_answer: 'true', question: 'question0'},
            {correct_answer: 'true', question: 'question1'},
            {correct_answer: 'false', question: 'question2'}
          ]} 
          answers={['true', 'false', 'true']} 
        />
      </MemoryRouter>
    )).toMatchSnapshot()
  })

  it('restarts quiz on reset click', () => {
    const mockHistory = getMockHistory()
    const resetMock = jest.fn()
    const instance = render(
      <MemoryRouter>
        <Results 
          finished
          reset={resetMock}
          history={mockHistory}
          questions={[
            {correct_answer: 'true', question: 'question0'},
            {correct_answer: 'true', question: 'question1'},
            {correct_answer: 'false', question: 'question2'}
          ]} 
          answers={['true', 'false', 'true']} 
        />
      </MemoryRouter>
    )
    instance.root.findByType(Button).props.onClick()
    expect(mockHistory.replace).toHaveBeenCalledWith('/')
    expect(resetMock).toHaveBeenCalled()
  })
})
