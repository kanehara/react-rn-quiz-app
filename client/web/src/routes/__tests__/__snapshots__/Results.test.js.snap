// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results redirects to / if not finished 1`] = `
<div
  data-testid="homeRedirect"
/>
`;

exports[`Results redirects to /quiz if inProgress 1`] = `
<div
  data-testid="quizRedirect"
/>
`;

exports[`Results renders results 1`] = `
<div
  data-testid="homeRedirect"
/>
`;
